compiler/luci/export/src/ExportMetadata.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ExportMetadata.h"
+
+namespace luci
+{
+
+void ExportMetadata::link_source(uint32_t node_id, uint32_t source_id)
+{
+  if (_op_table.find(node_id) == _op_table.end())
+    _op_table.emplace(node_id, std::set<uint32_t>());
+  _op_table.at(node_id).emplace(source_id);
+}
+
+void ExportMetadata::add_source(uint32_t source_id, std::string origin_name)
+{
+  if (_source_table.find(source_id) == _source_table.end())
+    _source_table.emplace(source_id, origin_name);
+  else
+  {
+    // NOTE Model with only one subgraph must not reach here.
+    //      However, model with multiple subgraph can reach here.
+    //      For now, as we do not consider about multiple subgraph in profiling,
+    //      just do nothing and support those cases in the future.
+  }
+}
+
+} // namespace luci
+
+namespace luci
+{
+
+const std::vector<uint8_t> encoded_source_table(ExportMetadata &metadata)
+{
+  std::vector<uint8_t> data;
+
+  const auto source_table = metadata.source_table();
+  for (auto &kv : source_table)
+  {
+    const auto id = kv.first;
+    data.emplace_back(0xFF & (id >> 0 * 8));
+    data.emplace_back(0xFF & (id >> 1 * 8));
+    data.emplace_back(0xFF & (id >> 2 * 8));
+    data.emplace_back(0xFF & (id >> 3 * 8));
```

Comment:
(Originally written by @seanshpark )

maybe consider something like this...
```cxx
union {
  uint32_t u32;
  uint8_t u8[4];
} u328_t;

u328_t u328;
u328.u32 = id;
for (...) data.emplace_back(u328.u8[i])
```

Reply:
if `data.emplace_back(u328.u8)` is possible then this is better :)

Reply:
I thunk your intention is to handle these similar codes effectively.
If it is right, I agree with the intention and thus I am thinking to introduce following function.
```cpp
void append(std::vector<uint8_t> &to, uint32_t value)
{
  to.emplace_back(0xFF & (value >> 0 * 8));
  to.emplace_back(0xFF & (value >> 1 * 8));
  to.emplace_back(0xFF & (value >> 2 * 8));
  to.emplace_back(0xFF & (value >> 3 * 8));
}
```

Reply:
> I thunk your intention is to handle these similar codes effectively.

Well, providing this as a function is one thing and this is necessary to reduce duplicate codes.
Actually how to make simple serializing 32 bit to four 8 bit is the main topic I like to solve :)